File: Editorial/0000-0100/0089_Gray_Code/Go/1_Backtracking.go
func grayCode(n int) []int {
    seen := map[int]bool{0: true}
    res := []int{0}
    var helper func(n int, seen map[int]bool) bool
    helper = func(n int, seen map[int]bool) bool {
        if len(res) == 1<<n {
            return true
        }
        curr := res[len(res)-1]
        for i := 0; i < n; i++ {
            next := curr ^ (1 << i)
            if seen[next] == false {
                seen[next] = true
                res = append(res, next)
                if helper(n, seen) {
                    return true
                }
                seen[next] = false
                res = res[:len(res)-1]
            }
        }
        return false
    }
    helper(n, seen)
    return res
}

File: Editorial/0000-0100/0089_Gray_Code/C/1_Backtracking.c
// Define the struct for uthash
typedef struct {
    int id;
    UT_hash_handle hh;
} Record;
int grayCodeHelper(int *res, int n, int *idx, Record **records);
int *grayCode(int n, int *returnSize) {
    *returnSize = 1 << n;
    int *res = malloc((*returnSize) * sizeof(*res));
    memset(res, 0, (*returnSize) * sizeof(*res));
    Record *records = NULL;
    Record *tmp = malloc(sizeof(*tmp));
    tmp->id = 0;
    HASH_ADD_INT(records, id, tmp);
    int start_idx = 1;
    grayCodeHelper(res, n, &start_idx, &records);
    // Clean up the hash table
    while (records) {
        tmp = records;
        HASH_DEL(records, tmp);
        free(tmp);
    }
    return res;
}
int grayCodeHelper(int *res, int n, int *idx, Record **records) {
    if (*idx == 1 << n) return 1;
    int current = res[*idx - 1];
    for (int i = 0; i < n; i++) {
        int next = current ^ (1 << i);
        Record *tmp;
        HASH_FIND_INT(*records, &next, tmp);
        if (tmp == NULL) {
            tmp = malloc(sizeof(*tmp));
            tmp->id = next;
            HASH_ADD_INT(*records, id, tmp);
            res[*idx] = next;
            (*idx)++;
            if (grayCodeHelper(res, n, idx, records)) return 1;
            HASH_DEL(*records, tmp);
            free(tmp);
            (*idx)--;
        }
    }
    return 0;
}
